osquery/logger/plugins/buffered.cpp

```diff
@@ -67,14 +74,73 @@ void BufferedLogForwarder::check() {
       VLOG(1) << "Error sending status to logger: " << status.getMessage();
     } else {
       // Clear the status logs once they were sent.
-      iterate(indexes, ([&results, this](std::string& index) {
+      iterate(indexes, ([this](std::string& index) {
                 if (!isStatusIndex(index)) {
                   return;
                 }
-                deleteDatabaseValue(kLogs, index);
+                std::lock_guard<std::recursive_mutex> lock(count_mutex_);
+                deleteValueWithCount(kLogs, index);
               }));
     }
   }
+
+  // Purge any logs exceeding the max after our send attempt
+  if (FLAGS_buffered_log_max > 0) {
+    purge();
+  }
+}
+
+void BufferedLogForwarder::purge() {
+  std::lock_guard<std::recursive_mutex> lock(count_mutex_);
+  size_t buffer_count = getBufferCount();
+  if (buffer_count <= FLAGS_buffered_log_max) {
+    return;
+  }
+
+  size_t purge_count = buffer_count - FLAGS_buffered_log_max;
+
+  // Collect purge_count indexes of each type (result/status) before
+  // partitioning to find the oldest. Note this assumes that the indexes are
+  // returned in ascending lexicographic order (true for RocksDB).
```

Comment:
Note this. Are we okay with that? If not, I think we'll have to do a full scan for the indexes if we want to make sure to throw away the oldest.